dwst/scripts/particles.js

```diff
@@ -20,7 +20,25 @@ const specialChars = [
   '\\',
 ];
 
-export class InvalidParticles extends Error { }
+export class InvalidParticles extends Error {
+  constructor(expression, errorPosition, expected, got) {
+    super();
+    this.expression = expression;
+    this.errorPosition = errorPosition;
+    this.expected = expected;
+    this.got = got;
+  }
+}
+
+export function invalidToMlog(invalidParticle) {
+  const padding = ' '.repeat(invalidParticle.errorPosition);
```

Comment:
The current code does not calculate errorPosition. However, this problem should become apparent by adding tests that test the code. Therefore, I would start by working on the tests cases and only fix errorPosition calculation after we have tests to verify correctness of the implementation.

Reply:
After letting this simmer for a while, I came to the conclusion that it would be probably easier to not store errorPosition as a number in the exception. If you were doing C programming you'd probably want to use a number because strings are a bit inconvenient in C but since dwst is written in javascript it would probably be easier to just store the valid part and the invalid part of the template in the exception as strings.

Reply:
For example parsing the invalid tamplate `'${foo(,456)}'` would result in `valid: '${foo(', invalid: ',456)}'` as exception keys.

Reply:
You could then simply do `const padding = ' '.repeat(invalidParticle.valid.length);`

Reply:
Perhaps storing this information as a number makes more sense in the end but you may still wish to use notation like `'${foo('.length` in test code to avoid magic numbers there.